File: Palace/AppInfrastructure/TPPConfiguration.m
#import "TPPConfiguration.h"
#import "TPPAppDelegate.h"

#import "UILabel+NYPLAppearanceAdditions.h"
#import "UIButton+NYPLAppearanceAdditions.h"
#import "Palace-Swift.h"

#if defined(FEATURE_DRM_CONNECTOR)
#import <ADEPT/ADEPT.h>
#endif

@implementation TPPConfiguration

+ (NSURL *)mainFeedURL
{
  NSURL *const customURL = [TPPSettings sharedSettings].customMainFeedURL;

  if(customURL) return customURL;

  NSURL *const accountURL = [TPPSettings sharedSettings].accountMainFeedURL;
  return accountURL;
}

+ (NSURL *)minimumVersionURL
{
  return [NSURL URLWithString:@"http://www.librarysimplified.org/simplye-client/minimum-version"];
}

+ (UIColor *)accentColor
{
  return [UIColor colorWithRed:0.0/255.0 green:144/255.0 blue:196/255.0 alpha:1.0];
}

+ (UIColor *)backgroundColor
{
  if (@available(iOS 13, *)) {
    return [UIColor colorNamed: @"ColorBackground"];
  }
  return [UIColor colorWithWhite:250/255.0 alpha:1.0];
}

+ (UIColor *)readerBackgroundColor
{
  return [UIColor colorWithWhite:250/255.0 alpha:1.0];
}

// OK to leave as static color because it's reader-only
+ (UIColor *)readerBackgroundDarkColor
{
  return [UIColor colorWithWhite:5/255.0 alpha:1.0];
}

// OK to leave as static color because it's reader-only
+ (UIColor *)readerBackgroundSepiaColor
{
  return [UIColor colorWithRed:250/255.0 green:244/255.0 blue:232/255.0 alpha:1.0];
}

// OK to leave as static color because it's reader-only
+ (UIColor *)backgroundMediaOverlayHighlightColor
{
  return [UIColor yellowColor];
}

// OK to leave as static color because it's reader-only
+ (UIColor *)backgroundMediaOverlayHighlightDarkColor
{
  return [UIColor orangeColor];
}

// OK to leave as static color because it's reader-only
+ (UIColor *)backgroundMediaOverlayHighlightSepiaColor
{
  return [UIColor yellowColor];
}

+ (NSString *)systemFontFamilyName
{
  return @"OpenSans";
}

+ (NSString *)systemFontName
{
  return @"OpenSans-Regular";
}

+ (NSString *)semiBoldSystemFontName
{
  return @"OpenSans-SemiBold";
}

+ (NSString *)boldSystemFontName
{
  return @"OpenSans-Bold";
}

+ (CGFloat)defaultTOCRowHeight
{
  return 56;
}

+ (CGFloat)defaultBookmarkRowHeight
{
  return 100;
}

+ (UINavigationBarAppearance *)defaultAppearance
{
  return [TPPConfiguration appearanceWithBackgroundColor:[TPPConfiguration backgroundColor]];
}

+ (UINavigationBarAppearance *)appearanceWithBackgroundColor:(UIColor *)backgroundColor
{
  UINavigationBarAppearance *appearance = [[UINavigationBarAppearance alloc] init];
  [appearance configureWithOpaqueBackground];
  [appearance setBackgroundColor: backgroundColor];
  [appearance setTitleTextAttributes:@{NSFontAttributeName: [UIFont semiBoldPalaceFontOfSize:18.0]}];
  return appearance;
}

@end

File: Palace/AppInfrastructure/TPPConfiguration.swift
import UIKit

@objc class TPPConfiguration: NSObject {

  @objc static func mainFeedURL() -> URL? {
    if let customURL = TPPSettings.shared.customMainFeedURL {
      return customURL
    }
    return TPPSettings.shared.accountMainFeedURL
  }

  @objc static func minimumVersionURL() -> URL? {
    return URL(string: "http://www.librarysimplified.org/simplye-client/minimum-version")
  }

  @objc static func accentColor() -> UIColor {
    return UIColor(red: 0.0/255.0, green: 144.0/255.0, blue: 196.0/255.0, alpha: 1.0)
  }

  @objc static func backgroundColor() -> UIColor {
    return UIColor(named: "ColorBackground") ?? UIColor(white: 250.0/255.0, alpha: 1.0)
  }

  @objc static func readerBackgroundColor() -> UIColor {
    return UIColor(white: 250.0/255.0, alpha: 1.0)
  }

  @objc static func readerBackgroundDarkColor() -> UIColor {
    return UIColor(white: 5.0/255.0, alpha: 1.0)
  }

  @objc static func readerBackgroundSepiaColor() -> UIColor {
    return UIColor(red: 250.0/255.0, green: 244.0/255.0, blue: 232.0/255.0, alpha: 1.0)
  }

  @objc static func backgroundMediaOverlayHighlightColor() -> UIColor {
    return .yellow
  }

  @objc static func backgroundMediaOverlayHighlightDarkColor() -> UIColor {
    return .orange
  }

  @objc static func backgroundMediaOverlayHighlightSepiaColor() -> UIColor {
    return .yellow
  }

  @objc static func systemFontFamilyName() -> String {
    return "OpenSans"
  }

  @objc static func systemFontName() -> String {
    return "OpenSans-Regular"
  }

  @objc static func semiBoldSystemFontName() -> String {
    return "OpenSans-SemiBold"
  }

  @objc static func boldSystemFontName() -> String {
    return "OpenSans-Bold"
  }

  @objc static func defaultTOCRowHeight() -> CGFloat {
    return 56
  }

  @objc static func defaultBookmarkRowHeight() -> CGFloat {
    return 100
  }

  @objc static func defaultAppearance() -> UINavigationBarAppearance {
    return appearance(withBackgroundColor: backgroundColor())
  }

  @objc static func appearance(withBackgroundColor backgroundColor: UIColor) -> UINavigationBarAppearance {
    let appearance = UINavigationBarAppearance()
    appearance.configureWithOpaqueBackground()
    appearance.backgroundColor = backgroundColor
    appearance.titleTextAttributes = [
      .font: UIFont.semiBoldPalaceFont(ofSize: 18.0)
    ]
    return appearance
  }
}
